sql/src/main/java/io/crate/protocols/postgres/ConnectionContext.java

```diff
@@ -344,8 +357,50 @@ public void channelDisconnected(ChannelHandlerContext ctx, ChannelStateEvent e)
 
     private void handleStartupBody(ChannelBuffer buffer, Channel channel) {
         session = readStartupMessage(buffer);
-        Messages.sendAuthenticationOK(channel);
+        if (authService != null && authService.enabled()) {
+            authenticate(channel);
+        } else {
+            Messages.sendAuthenticationOK(channel);
+            sendReadyForQuery(channel);
+        }
+    }
+
+    private void authenticate(Channel channel) {
+        SessionContext ctx = session.sessionContext();
+        InetAddress address = ((InetSocketAddress) channel.getRemoteAddress()).getAddress();
+        AuthenticationMethod authMethod = authService.resolveAuthenticationType(ctx.userName(), address);
+        if (authMethod == null) {
+            String errorMessage = String.format(
+                Locale.ENGLISH,
+                "No valid auth.host_based entry found for host \"%s\", user \"%s\", schema \"%s\"",
+                address.getHostAddress(), ctx.userName(), ctx.defaultSchema()
+            );
+            Messages.sendAuthenticationError(channel, errorMessage);
+        } else {
+            authMethod.pgAuthenticate(channel, ctx)
+                .whenComplete(new AuthorizationListener() {
+                    @Override
+                    public void onAuthorization(Boolean success) {
+                        if (success) {
+                            if (LOGGER.isTraceEnabled()) {
+                                LOGGER.error("Authentication succeeded user \"{}\" and method \"{}\".",
```

Comment:
the implementation of the authentication method already sends the error response to the client which then terminates the connection.